share/spack/win-compiler-wrapper/cl.cxx

```diff
@@ -0,0 +1,440 @@
+    {
+        this->libDirArgs.push_back(this->composeLibPathArg(libDir));
+    }
+}
+
+std::string ToolChainInvocation::composeCLI() {
+    std::string SpackCompilerCLI;
+    auto addToCLI = [&](StrList args){
+        for(auto& arg: args){
+            SpackCompilerCLI += arg + " ";
+        }
+    };
+    addToCLI(this->CommandArgs);
+    addToCLI(this->includeArgs);
+    addToCLI(this->libDirArgs);
+    addToCLI(this->libArgs);
+    return SpackCompilerCLI;
+}
+
+void ToolChainInvocation::cleanupHandles() {
+
+    this->safeHandleCleanup(this->procInfo.hProcess);
+    this->safeHandleCleanup(this->procInfo.hThread);
+}
+
+void ToolChainInvocation::safeHandleCleanup(HANDLE &handle) {
+    if (!CloseHandle(handle)) {
+        std::stringstream os_error;
+        os_error << GetLastError();
+        throw SpackException(os_error.str());
+    }
+}
+
+void ClInvocation::loadToolchainDependentSpackVars(SpackEnvState &spackenv) {
+    this->spackCommand = spackenv.SpackCC;
+}
+
+void LdInvocation::loadToolchainDependentSpackVars(SpackEnvState &spackenv) {
+    this->spackCommand = std::string("link.exe");
+}
+
+void IntelFortranInvocation::loadToolchainDependentSpackVars(SpackEnvState &spackenv) {
+    this->spackCommand = spackenv.SpackFC;
+}
+
+
+std::unique_ptr<ToolChainInvocation> ToolChainFactory::ParseToolChain(int argc, char const* const * argv) {
+    std::string command(argv[0]);
+    char const* const* cli(++argv);
+    stripPathandExe(command);
+    try {
+        std::string language = SupportedTools.at(command);
+        std::unique_ptr<ToolChainInvocation> Tool;
+        if(language == "c/c++") {
+            Tool = std::unique_ptr<ClInvocation>(new ClInvocation(command, cli));
+        }
+        else if(language == "intel fortran") {
+            Tool = std::unique_ptr<IntelFortranInvocation>(new IntelFortranInvocation(command, cli));
+        }
+        else {
+            // If it's not c/c++ or fortran we're linking
+            Tool = std::unique_ptr<LdInvocation>(new LdInvocation(command, cli));
+        }
+        return Tool;
+    } catch (std::out_of_range& e) {
+        throw SpackUnknownCompilerException(command);
+    }
+}
+
+void ToolChainFactory::stripPathandExe(std::string &command) {
+    stripPath(command);
+    stripExe(command);
+};
+void ToolChainFactory::stripExe(std::string &command) {
+    std::string::size_type loc = command.rfind(".exe");
+    if (std::string::npos != loc)
+        command.erase(loc);
+}
+
+void ToolChainFactory::stripPath(std::string &command) {
+    command.erase(0, command.find_last_of("\\/") + 1);
+}
+
+const std::map<std::string, std::string> ToolChainFactory::SupportedTools{
+                                                        {"cl", "c/c++"},
+                                                        {"link", "link"},
+                                                        {"ifort", "intel fortran"},
+                                                        {"ifx", "intel fortran"}
```

Comment:
IMO using an enum would be cleaner for values.

Also, should we be mapping `ifort` to `fortran` vs. `intel-fortran`? I assume we'll want to support other fortran compilers in the future (that's not something that needs to change here, although I'm curious what the consequences are of making it generic like that here).

Reply:
I'll switch to the enum. 

> should we be mapping ifort to `fortran` vs. `intel-fortran`
I am of the opinion that since `ifort` is unique to the intel fortran environment, which like MSVC, generally likes to do its own thing there is value in making the distinction between `intel fortran` and vanilla `fortran` particularly if some specific flags or handling becomes necessary down the line and we need to be able to distinguish between a ported `gfortran` and `ifx`.